diffkemp/simpll/diffkemp_patterns.h

```diff
@@ -0,0 +1,183 @@
+ * differently named arguments in each version of the pattern. However, it is
+ * still necessary that the signatures match.
+ *
+ * To define a pattern that ends with a resolution of a condition, use
+ * CONDITION_PATTERN_OLD and CONDITION_PATTERN_NEW macros. The pattern should
+ * return a boolean value, used as the condition. It is not necessary to declare
+ * output mapping for the condition variable. See `condition_only.c` pattern in
+ * the aforementioned example folder.
+ *
+ * To define a value pattern, defining a semantic equivalence between two
+ * values, use VALUE_PATTERN macro. First macro argument is the function
+ * name, the second and the third are the old and the new value, respectively.
+ * When using extern global variables, use pointer to the value instead.
+ *
+ * Examples of a value pattern:
+ *  VALUE_PATTERN(value, 0b110UL << 8, 0b101UL << 7);
+ *  VALUE_PATTERN(global_value, 30, &extern_var);
+ *
+ * Patterns defined in this way can then be used by the diffkemp tool using
+ * standard -p flag, in the same way as the LLVM patterns. The compiled .ll
+ * pattern file will be located in the same location as the .c pattern file from
+ * which it was compiled. It is also possible to purely compile the .c pattern
+ * file to .ll file without performing comparison, by using the compile-pattern
+ * sub-command.
+ *
+ * When writting patterns for kernel, it is also necessary to provide following
+ * definitions and includes at the very beginning of the file, before including
+ * other kernel headers:
+ *  #define __KERNEL__
+ *  #define __BPF_TRACING__
+ *  #define __HAVE_BUILTIN_BSWAP16__
+ *  #define __HAVE_BUILTIN_BSWAP32__
+ *  #define __HAVE_BUILTIN_BSWAP64__
+ *  #include <linux/kconfig.h>
+ * Then, following include paths in the following order must be provided to the
+ * compiler:
+ *  -I{linux}/arch/x86/include/
+ *  -I{linux}/arch/x86/include/generated/
+ *  -I{linux}/include/
+ *  -I{linux}/arch/x86/include/uapi
+ *  -I{linux}/arch/x86/include/generated/uapi
+ *  -I{linux}/include/uapi
+ *  -I{linux}/include/generated/uapi
+ * This can be done automatically by the pattern compiler by providing the path
+ * to the kernel source files using --c-pattern-kernel-path option.
+ *
+ * Patterns written in C can also be loaded from YAML file, in the same way as
+ * the LLVM patterns. The YAML file must contain field `patterns` with the list
+ * of pattern files. Additionally, it is possible to provide extra clang options
+ * for each individual pattern, using field `clang_append`, by providing a map
+ * of pattern names to lists of clang options to append to them. For examples,
+ * see the tests/regression/custom_patterns/c/
+ */
+
+#ifndef DIFFKEMP_SIMPLL_DIFFKEMP_PATTERNS_H
+#define DIFFKEMP_SIMPLL_DIFFKEMP_PATTERNS_H
+
+// Internal definitions
+
+#define __DIFFKEMP_STRINGIFY_IMPL(macro) #macro
+#define __DIFFKEMP_STRINGIFY(macro) __DIFFKEMP_STRINGIFY_IMPL(macro)
```

Comment:
The additional layer of indirection is necessary to properly expand the arguments, for example if a macro itself is passed. Without it, it will stringify the macro name, not the macro value.

Reply:
Makes sense, thanks.